### PR TITLE
Revert "mass-replace deprecated usage of find -perm (Fixes #9044)"

### LINKED ISSUE #9044: find -perm /mode doesn't work on Darwin
nix works on cygwin but nixpkgs use a deprecated syntax which does not work with the latest cygwin anymore.

-perm +0100 should be -perm /0100

EDIT: after doing that change, Darwin has problems, likely due to using some ancient version of findutils.

### PR DESCRIPTION
**Pull request by dredozubov**
This makes find unusable on darwin, breaking most of the builds.

This reverts commit 862900e42e00dbc83215def6df1091f5129a8930.

see #9044

[Review on Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/9603)

### CODE REVIEW

**Comment by vcunat** [comment]
I rather thought of "reverting only for Darwin", and for simplicity, only in the necessary cases (to minimize file splits). But if "most people" prefer this simple revert, go ahead. (I only fear that noone will really fix this, so we'll stay with broken Cygwin and using deprecated style besides.)

**Comment by dredozubov** [comment]
Not sure why the build is broken after a simple revert. Is something broken on travis maybe?

**Comment by vcunat** [comment]
Travis doesn't work for huge rebuilds, at least this free (as beer, IIRC) version doesn't. Someone with Darwin should actually test the thing works before merging, whether we go for this approach or another.

**Comment by zimbatm** [comment]
@vcunat what version of findutils is cygwin using ? I'm surprised `-perm +` is not supported since it's part of POSIX while `-perm /` is not.

**Comment by vcunat** [comment]
I don't know, I just trusted @tohl on the cygwin point.

**Comment by usharf** [comment]
How do I work around this right now? I can't compile anything depending on `zlib` on Darwin at the moment.

```sh
...
strip: can't process non-object and non-archive file: /nix/store/ry1xfxg5rm02caqzszjidqrgvzhp75hg-zlib-1.2.8/lib/pkgconfig/zlib.pc
patching script interpreter paths in /nix/store/ry1xfxg5rm02caqzszjidqrgvzhp75hg-zlib-1.2.8
find: -perm: /0100: illegal mode string
builder for ‘/nix/store/ydby11y8c66mip7c0hrwba1piipl2r91-zlib-1.2.8.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/h0prdxs8k2irmjbf0cxg4bv1sgcp65x7-curl-7.43.0.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/59sdl75fbpa0lr14k9zkyd810rgc6viv-sed-4.2.2.tar.bz2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/y4753yvsd0ya2b6wy0g5k3wn4lwfs533-xz-5.2.1.tar.bz2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/plsbi4kds37f6gn48gvnx5p6ick8izik-gnused-4.2.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/4riavg2sz9yvvbvwm80bjfd5kaqaznmh-xz-5.2.1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/46lrmvbirqjl5gb869sphvj6whmylfn6-stdenv-darwin-boot-2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/vnpirirkrc4958nhz1bsrwa5zzq536l4-bash-4.3-p39.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/vnpirirkrc4958nhz1bsrwa5zzq536l4-bash-4.3-p39.drv’ failed
/Users/uri/.nix-profile/bin/nix-shell: failed to build all dependencies
```

**Comment by vcunat** [comment]
@usharf: right now you can revert locally, for example ;-)

**Comment by usharf** [comment]
@vcunat do you mean I should checkout all the relevant files touched by that commit locally ...?

**Comment by vcunat** [comment]
@usharf: checkout the whole nixpkgs tree, of course.

**Comment by usharf** [comment]
@vcunat ok reverted to that commit... is there a way to make nix still use the cache or will that invalidate everything dependent?

**Comment by dredozubov** [comment]
@usharf afair no, there isn't. And this is a perfect argument for merging this asap. Currently you can't make user experience on darwin worse that it is.

**Comment by zimbatm** [comment]
Here is a little recap of everything (with some new stuff):

GOAL: We want a cross-platform way of finding executable files in bash scripts that can also be used during bootstrap. Unfortunately the various `find` implementations don't agree on their arguments.

BSD find (and thus OSX), busybox find and GNU findutils v4.4.2 (latest stable release and current NixOS version) all support `-type f -perm +0100`.

GNU findutils v4.4.2's man page says that `-perm +...` is deprecated and `-perm /...` should be used instead.

Cygwin is distribution findutils [versions 4.5.11 and 4.5.12](http://mirror.pkill.info/cygwin/x86_64/release/findutils/) which are [alpha releases](ftp://alpha.gnu.org/findutils/). These version have removed the `-perm +...` flag entirely. 

findutils is the only implementation which supports the `-executable` flag which does the same thing as `-perm +0100`.

The POSIX standard specifies [neither the `-perm +...` nor the `-perm /` and `-executable` options](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/find.html) (unlike what I said previously) which probably explains the incompatibilities.

So in short there is no easy way to be cross-platform. I would suggest to revert for now since Darwin has more users and then provide a `find-executable` shim that works in a cross-platform way.

Even more info: https://stackoverflow.com/questions/4458120/unix-find-search-for-executable-files#answer-29039160

**Comment by edolstra** [comment]
No need for a shim - just pass the right flags based on the platform. See e.g. `xargsFlags` and `stripAllFlags` (https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/strip.sh#L33) which are overriden on Darwin (https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/darwin/default.nix#L50).

**Comment by chamini2** [comment]
I tried this and it works, I vote for merging while a cross-platform solution is worked on.

**Comment by johbo** [comment]
Just also ran into this, like the idea with the `stripAllFlags` example, anybody preparing this already?

**Comment by vcunat** [comment]
Closing in favor of #9667 or its modifications if need be.

It wouldn't be good to merge a mass-rebuild change and the fix it up a few days later, as the resources spent by Hydra on it are large.
